Fix arithmetic assignment operations for multidimensional indexed arrays

hyenias opened the pull request:
This PR corrects #168 for indexed arrays having more than one level.
Turns out ksh was only keeping track of the subscript number for
assignment in lvalue's nosub variable. By saving the actual subscript
reference, the result can be assigned to its proper destination instead
of putting the result into the last looked value or subscript location.

src/cmd/ksh93/include/streval.h: struct lval:
- Create a new pointer named sub to hold the reference that nosub describes.

src/cmd/ksh93/sh/arith.c: arith():
- Adjust LOOKUP: for lvalue ARITH_ASSIGNOP operations on indexed arrays to
  save the np of the destination subscript for later use.
- Adjust ASSIGN: to act when lvalue's nosub > 0 which happens as the last
  step in the arithmetic parsing loop for assignment operations. Only indexed
  arrays will have a nosub value > 0. All others have a nosub of 0 unless they
  are involved in a unary operation (`++, --`) which sets nosub to -1. All said
  in the context of assignment operations like `(( arr[0][1] += 1 ))`.

src/cmd/ksh93/sh/streval.c:
- Initialize the new sub pointer to 0.

hyenias committed:
Fix arithmetic assignment operations for multidimensional indexed arrays
This PR corrects #168 for indexed arrays having more than one level.
Turns out ksh was only keeping track of the subscript number for
assignment in lvalue's nosub variable. By saving the actual subscript
reference, the result can be assigned to its proper destination instead
of putting the result into the last looked value or subscript location.

src/cmd/ksh93/include/streval.h: struct lval:
- Create a new pointer named sub to hold the reference that nosub describes.

src/cmd/ksh93/sh/arith.c: arith():
- Adjust LOOKUP: for lvalue ARITH_ASSIGNOP operations on indexed arrays to
  save the np of the destination subscript for later use.
- Adjust ASSIGN: to act when lvalue's nosub > 0 which happens as the last
  step in the arithmetic parsing loop for assignment operations. Only indexed
  arrays will have a nosub value > 0. All others have a nosub of 0 unless they
  are involved in a unary operation (++, --) which sets nosub to -1. All said
  in the context of assignment operations like (( arr[0][1] += 1 )).

src/cmd/ksh93/sh/streval.c:
- Initialize the new sub pointer to 0.

McDutchie commented:
Great work on the fix. It would be good to have some regression tests to go with it, though. They don't have to be as extensive as your impressive [arrmath.ksh](https://github.com/hyenias/code/blob/main/ksh/arrmath.ksh) script (which has an incompatible licence anyway) but we do need something to ensure these bugs don't return.

hyenias commented:
Will do. I will work on some regression tests. Just wanted to get my code changes out for all to review. Need to create a NEWS entry as well.

hyenias committed:
Add in a few tests to identify multilevel subscripted indexed array issues.

src/cmd/ksh93/tests/arrays2.sh:
- Created a few multidimensional indexed array tests for assignment operations
  like += as an example.

hyenias committed:
Merge branch 'master' into arith-array

McDutchie commented:
The fix looks good to me. I don't have any deeper understanding of the arith.c code at this point. It looks like you're far ahead of me there. So I don't really feel qualified to judge if this is the best possible fix or not. But it looks logical, especially when I trace it with some debug commands (for a command like `((arr[1][2][3][4]+=1))`, the parent node used is `arr[1][2][3]` which makes sense), and your exhaustive arrmath.ksh script convincingly shows it works. I'm impressed!
